File: services/gateway/gateway-api/src/main/java/com/example/gateway/api/controller/TradeLogController.java
package com.example.gateway.api.controller;

import com.example.gateway.api.exception.ExceptionCode;
import com.example.gateway.api.exception.GatewayApiException;
import com.example.gateway.api.model.DividendGWModel;
import com.example.gateway.api.model.OptionJournalGWModel;
import com.example.gateway.api.model.ShareJournalGWModel;
import com.example.gateway.api.model.TradeLogGWModel;
import com.example.gateway.api.model.TradeSummaryGWModel;
import com.example.gateway.api.model.TransactionSettingsGWModel;
import com.example.gateway.api.service.TradeLogService;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.concurrent.ExecutionException;

@RestController
@RequestMapping(value = "/api/v1/tradelog", produces = "application/json")
public class TradeLogController {

    private TradeLogService tradeLogService;

    public TradeLogController(TradeLogService tradeLogService) {
        this.tradeLogService = tradeLogService;
    }


    @RequestMapping(value = "/all/{symbol}", method = RequestMethod.GET)
    @ResponseStatus(HttpStatus.OK)
    public TradeLogGWModel getAllBySymbol(
            @RequestHeader(name = "accountId") String accountId,
            @PathVariable(name = "symbol") String symbol) throws ExecutionException, InterruptedException {

        //todo validate input

        return tradeLogService.getAllBySymbol(accountId, symbol);
    }


    @RequestMapping(value = "/shares", method = RequestMethod.POST)
    @ResponseStatus(HttpStatus.OK)
    public ShareJournalGWModel createNewShareTrade(
            @RequestHeader(name = "accountId") String accountId,
            @RequestBody ShareJournalGWModel model) {
        return tradeLogService.createShareTrade(accountId, model);
    }


    @RequestMapping(value = "/shares/{id}", method = RequestMethod.DELETE)
    @ResponseStatus(HttpStatus.OK)
    public void deleteShareTrade(
            @RequestHeader(name = "accountId") String accountId,
            @PathVariable(name = "id") String transactionId) {
        tradeLogService.deleteShareTrade(accountId, transactionId);
    }


    @RequestMapping(value = "/options", method = RequestMethod.POST)
    @ResponseStatus(HttpStatus.OK)
    public OptionJournalGWModel createNewOptionTrade(
            @RequestHeader(name = "accountId") String accountId,
            @RequestBody OptionJournalGWModel model) {
        return tradeLogService.createOptionTrade(accountId, model);
    }


    @RequestMapping(value = "/options/{id}", method = RequestMethod.DELETE)
    @ResponseStatus(HttpStatus.OK)
    public void deleteOptionTrade(
            @RequestHeader(name = "accountId") String accountId,
            @PathVariable(name = "id") String transactionId) {
        tradeLogService.deleteOptionTrade(accountId, transactionId);
    }


    @RequestMapping(value = "/dividends", method = RequestMethod.POST)
    @ResponseStatus(HttpStatus.OK)
    public DividendGWModel createDividendRecord(
            @RequestHeader(name = "accountId") String accountId,
            @RequestBody DividendGWModel model) {
        return tradeLogService.createDividendRecord(accountId, model);
    }


    @RequestMapping(value = "/dividends/{id}", method = RequestMethod.DELETE)
    @ResponseStatus(HttpStatus.OK)
    public void deleteDividendRecord(
            @RequestHeader(name = "accountId") String accountId,
            @PathVariable(name = "id") String transactionId) {
        tradeLogService.deleteDividendRecord(accountId, transactionId);
    }


    @RequestMapping(value = "/symbols", method = RequestMethod.GET)
    @ResponseStatus(HttpStatus.OK)
    public List<String> getAllTradedSymbols(@RequestHeader(name = "accountId") String accountId) throws GatewayApiException {
        //todo validate input

        List<String> tradedSymbols = tradeLogService.getAllTradedSymbols(accountId);

        if (tradedSymbols == null || tradedSymbols.isEmpty()) {
            throw new GatewayApiException(ExceptionCode.TRADEJOURNAL_NO_SYMBOLS);
        }

        return tradedSymbols;
    }

    @RequestMapping(value = "/summary", method = RequestMethod.GET)
    @ResponseStatus(HttpStatus.OK)
    public List<TradeSummaryGWModel> getSummary(@RequestHeader("accountId") String accountId) throws GatewayApiException {
        List<TradeSummaryGWModel> summaryModels = tradeLogService.getSummary(accountId);

        if (summaryModels == null || summaryModels.isEmpty()) {
            throw new GatewayApiException(ExceptionCode.TRADEJOURNAL_NO_SYMBOLS);
        }

        return summaryModels;
    }

    @RequestMapping(value = "/transaction/settings/bulk", method = RequestMethod.PUT)
    @ResponseStatus(HttpStatus.OK)
    public void updateBulkSettings(@RequestHeader("accountId") String accountId,
                                   @RequestBody List<TransactionSettingsGWModel> modelList) {
        tradeLogService.updateGroupSettings(accountId, modelList);
    }

    @RequestMapping(value = "/transaction/settings/{id}", method = RequestMethod.PUT)
    @ResponseStatus(HttpStatus.OK)
    public void updateSettings(@RequestHeader("accountId") String accountId,
                               @PathVariable(name = "id") String transactionId,
                               @RequestBody TransactionSettingsGWModel model) {
        tradeLogService.updateGroupSetting(accountId, transactionId, model);
    }
}

File: services/gateway/gateway-api/src/main/kotlin/com/example/gateway/api/rest/controller/TradeLogController.kt
package com.example.gateway.api.rest.controller

import arrow.core.getOrHandle
import com.example.gateway.api.core.service.TradeLogService
import com.example.gateway.api.core.service.TransactionService
import com.example.gateway.api.rest.controller.TradeLogController.Companion.PROTOBUF_MEDIA_TYPE_VALUE
import com.example.gateway.api.rest.converter.*
import com.example.gateway.api.spec.model.*
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.web.bind.annotation.*
import java.util.*
import java.util.stream.Collectors

@RestController
@RequestMapping(value = ["/api/v1/tradelog"], produces = [PROTOBUF_MEDIA_TYPE_VALUE, MediaType.APPLICATION_JSON_VALUE])
class TradeLogController(private val tradeLogService: TradeLogService,
                         private val transactionService: TransactionService): SecureController {

    companion object {
        const val PROTOBUF_MEDIA_TYPE_VALUE = "application/x-protobuf"
    }


    @RequestMapping(value = ["/all/{symbol}", "/all/{symbol}/"], method = [RequestMethod.GET])
    @ResponseStatus(HttpStatus.OK)
    fun getAll(
            @RequestHeader(value = "x-auth-key", required = true) token: String,
            @RequestHeader(name = "portfolio-id", required = true) portfolioId: UUID,
            @PathVariable(name = "symbol", required = true) symbol: String): GWTradeLogDto {

        //todo: validate input
        val support = getSupportData(token, portfolioId)
        val tradeLogModel = tradeLogService.getAll(support, symbol)

        return TradeLogConverter.toDto(tradeLogModel)
    }

    @RequestMapping(value = ["/summary", "/summary/"], method = [RequestMethod.GET])
    @ResponseStatus(HttpStatus.OK)
    fun getSummary(
            @RequestHeader(value = "x-auth-key", required = true) token: String,
            @RequestHeader(name = "portfolio-id", required = true) portfolioId: UUID
    ): GWTradeSummaryResponse {
        val support = getSupportData(token, portfolioId)
        val summaryModels = tradeLogService.getSummary(support)

        val dtos = summaryModels
                .getOrHandle { throw it }
                .stream().map { TradeSummaryConverter.toDto(it) }
                .collect(Collectors.toList())

        return GWTradeSummaryResponse.newBuilder()
                .addAllItems(dtos)
                .build()
    }

    @RequestMapping(value = ["/summary/matrix", "/summary/matrix/"], method = [RequestMethod.GET])
    @ResponseStatus(HttpStatus.OK)
    fun getSummaryMatrix(@RequestHeader(value = "x-auth-key", required = true) token: String,
                         @RequestHeader(name = "portfolio-id", required = true) portfolioId: UUID,
                         @RequestParam(name = "shares-only", required = false, defaultValue = "false") sharesOnly: Boolean

    ): GWSummaryMatrixResponse {
        val support = getSupportData(token, portfolioId)
        val summaryModels = transactionService
                .getSummaryMatrix(support, sharesOnly)
                .getOrHandle { throw it }

        val mapDto = SummaryMatrixConverter.toGWDto(summaryModels)

        return GWSummaryMatrixResponse.newBuilder()
                .putAllYears(mapDto)
                .build()
    }


    @RequestMapping(value = ["/shares", "/shares/"], method = [RequestMethod.POST])
    @ResponseStatus(HttpStatus.OK)
    fun createShareTransaction(
            @RequestHeader(value = "x-auth-key", required = true) token: String,
            @RequestHeader(name = "portfolio-id", required = true) portfolioId: UUID,
            @RequestBody dto: GWCreateShareJournalDto): GWShareJournalDto {

        //todo: validate input
        val support = getSupportData(token, portfolioId)
        val createModel = CreateShareJournalConverter.toModel(support, dto)
        val createdModel = tradeLogService
                .createShareTransaction(support, createModel)
                .getOrHandle { throw it }

        return ShareJournalConverter.toGWDto(createdModel)
    }

    @RequestMapping(value = ["/shares/{transactionId}", "/shares/{transactionId}/"], method = [RequestMethod.PUT])
    @ResponseStatus(HttpStatus.OK)
    fun editShareTransaction(
            @RequestHeader(value = "x-auth-key", required = true) token: String,
            @RequestHeader(name = "portfolio-id", required = true) portfolioId: UUID,
            @PathVariable(name = "transactionId", required = true) transactionId: String,
            @RequestBody dto: GWShareJournalDto) {

        //todo: validate input
        val support = getSupportData(token, portfolioId)
        val updateModel = ShareJournalConverter.toModel(support, dto)
        tradeLogService
                .editShareTransaction(support, updateModel)
                .getOrHandle { throw it }
    }

    @RequestMapping(value = ["/shares/{transactionId}", "/shares/{transactionId}/"], method = [RequestMethod.DELETE])
    @ResponseStatus(HttpStatus.OK)
    fun deleteShareTransaction(
            @RequestHeader(value = "x-auth-key", required = true) token: String,
            @RequestHeader(name = "portfolio-id", required = true) portfolioId: UUID,
            @PathVariable(name = "transactionId", required = true) transactionId: String) {

        //todo: validate input
        val support = getSupportData(token, portfolioId)
        tradeLogService
                .deleteShareTransaction(support, transactionId)
                .getOrHandle { throw it }
    }

    @RequestMapping(value = ["/options", "/options/"], method = [RequestMethod.POST])
    @ResponseStatus(HttpStatus.OK)
    fun createOptionTransaction(
            @RequestHeader(value = "x-auth-key", required = true) token: String,
            @RequestHeader(name = "portfolio-id", required = true) portfolioId: UUID,
            @RequestBody dto: GWCreateOptionJournalDto): GWOptionJournalDto {

        //todo: validate input
        val support = getSupportData(token, portfolioId)
        val createModel = CreateOptionJournalConverter.toModel(support, dto)
        val createdModel = tradeLogService
                .createOptionTransaction(support, createModel)
                .getOrHandle { throw it }

        return OptionJournalConverter.toGWDto(createdModel)
    }

    @RequestMapping(value = ["/options/{transactionId}", "/options/{transactionId}/"], method = [RequestMethod.PUT])
    @ResponseStatus(HttpStatus.OK)
    fun editOptionTransaction(
            @RequestHeader(value = "x-auth-key", required = true) token: String,
            @RequestHeader(name = "portfolio-id", required = true) portfolioId: UUID,
            @PathVariable(name = "transactionId", required = true) transactionId: String,
            @RequestBody dto: GWOptionJournalDto) {

        //todo: validate input
        val support = getSupportData(token, portfolioId)
        val updateModel = OptionJournalConverter.toModel(support, dto)
        tradeLogService
                .editOptionTransaction(support, updateModel)
                .getOrHandle { throw it }
    }

    @RequestMapping(value = ["/options/{transactionId}", "/options/{transactionId}/"], method = [RequestMethod.DELETE])
    @ResponseStatus(HttpStatus.OK)
    fun deleteOptionTransaction(
            @RequestHeader(value = "x-auth-key", required = true) token: String,
            @RequestHeader(name = "portfolio-id", required = true) portfolioId: UUID,
            @PathVariable(name = "transactionId", required = true) transactionId: String) {

        //todo: validate input
        val support = getSupportData(token, portfolioId)
        tradeLogService
                .deleteOptionTransaction(support, transactionId)
                .getOrHandle { throw it }
    }

    @RequestMapping(value = ["/dividends", "/dividends/"], method = [RequestMethod.POST])
    @ResponseStatus(HttpStatus.OK)
    fun createDividendTransaction(
            @RequestHeader(value = "x-auth-key", required = true) token: String,
            @RequestHeader(name = "portfolio-id", required = true) portfolioId: UUID,
            @RequestBody dto: GWCreateDividendJournalDto): GWDividendJournalDto {

        //todo: validate input
        val support = getSupportData(token, portfolioId)
        val createModel = CreateDividendJournalConverter.toModel(support, dto)
        val createdModel = tradeLogService
                .createDividendTransaction(support, createModel)
                .getOrHandle { throw it }

        return DividendJournalConverter.toGWDto(createdModel)
    }

    @RequestMapping(value = ["/dividends/{transactionId}", "/dividends/{transactionId}/"], method = [RequestMethod.DELETE])
    @ResponseStatus(HttpStatus.OK)
    fun deleteDividendTransaction(
            @RequestHeader(value = "x-auth-key", required = true) token: String,
            @RequestHeader(name = "portfolio-id", required = true) portfolioId: UUID,
            @PathVariable(name = "transactionId", required = true) transactionId: String) {

        //todo: validate input
        val support = getSupportData(token, portfolioId)
        tradeLogService
                .deleteDividendTransaction(support, transactionId)
                .getOrHandle { throw it }
    }

    @RequestMapping(value = ["/settings/{transactionId}", "/settings/{transactionId}/"], method = [RequestMethod.PUT])
    @ResponseStatus(HttpStatus.OK)
    fun updateSetting(
            @RequestHeader(value = "x-auth-key", required = true) token: String,
            @RequestHeader(name = "portfolio-id", required = true) portfolioId: UUID,
            @PathVariable(name = "transactionId", required = true) transactionId: String,
            @RequestBody dto: GWTransactionSettingsDto) {

        //todo: validate input
        val support = getSupportData(token, portfolioId)
        val model = TransactionSettingsConverter.toModel(dto)
        transactionService
                .updateTransactionSetting(support, model)
                .getOrHandle { throw it }
    }

    @RequestMapping(value = ["/settings/bulk", "/settings/bulk/"], method = [RequestMethod.PUT])
    @ResponseStatus(HttpStatus.OK)
    fun updateSettings(
            @RequestHeader(value = "x-auth-key", required = true) token: String,
            @RequestHeader(name = "portfolio-id", required = true) portfolioId: UUID,
            @RequestBody dto: GWTransactionSettingsBulkDto) {

        //todo: validate input
        val support = getSupportData(token, portfolioId)
        val models = TransactionSettingsConverter.toModels(dto)
        transactionService
                .updateTransactionSettings(support, models)
                .getOrHandle { throw it }
    }
}